jsonrpc/server.go

```diff
@@ -211,59 +216,83 @@ func (s *Server) HandleReader(ctx context.Context, reader io.Reader) ([]byte, er
 		}
 	} else {
 		var batchReq []json.RawMessage
-		var batchRes []json.RawMessage
 
 		if batchJSONErr := dec.Decode(&batchReq); batchJSONErr != nil {
 			res.Error = Err(InvalidJSON, batchJSONErr.Error())
 		} else if len(batchReq) == 0 {
 			res.Error = Err(InvalidRequest, "empty batch")
 		} else {
-			for _, rawReq := range batchReq { // todo: handle async
-				var resObject *response
-
-				reqDec := json.NewDecoder(bytes.NewBuffer(rawReq))
-				reqDec.UseNumber()
-
-				req := new(request)
-				if jsonErr := reqDec.Decode(req); jsonErr != nil {
-					resObject = &response{
-						Version: "2.0",
-						Error:   Err(InvalidRequest, jsonErr.Error()),
-					}
-				} else {
-					var handleErr error
-					resObject, handleErr = s.handleRequest(ctx, req)
-					if handleErr != nil {
-						resObject = &response{
-							Version: "2.0",
-							Error:   Err(InvalidRequest, handleErr.Error()),
-						}
-						if !errors.Is(handleErr, ErrInvalidID) {
-							resObject.ID = req.ID
-						}
-					}
-				}
+			return s.handleBatchRequest(ctx, batchReq)
+		}
+	}
+
+	if res == nil {
+		return nil, nil
+	}
+	return json.Marshal(res)
+}
+
+func (s *Server) handleBatchRequest(ctx context.Context, batchReq []json.RawMessage) ([]byte, error) {
+	var (
+		responses []json.RawMessage
+		mutex     sync.Mutex
+	)
+
+	addResponse := func(response any) {
+		if responseJSON, err := json.Marshal(response); err != nil {
+			s.log.Errorw("failed to marshal response", "err", err)
+		} else {
+			mutex.Lock()
+			responses = append(responses, responseJSON)
+			mutex.Unlock()
+		}
+	}
 
-				if resObject != nil {
-					if resArr, jsonErr := json.Marshal(resObject); jsonErr != nil {
-						return nil, jsonErr
-					} else {
-						batchRes = append(batchRes, resArr)
-					}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var wg sync.WaitGroup
+	for _, rawReq := range batchReq {
+		reqDec := json.NewDecoder(bytes.NewBuffer(rawReq))
+		reqDec.UseNumber()
+
+		req := new(request)
+		if err := reqDec.Decode(req); err != nil {
+			addResponse(&response{
```

Comment:
Does the issue previously discussed still exist? That is: if one request has a decoding error, all requests for the batch should be canceled and only a single response sent back.

Reply:
There are two cases:
1. Entire request is an invalid JSON, example:
```
--> [
  {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},
  {"jsonrpc": "2.0", "method"
]
<-- {"jsonrpc": "2.0", "error": {"code": -32700, "message": "Parse error"}, "id": null}
```
2. Inside an array (in batch request) one or more items are not valid according to spec (but it's still valid JSON item):
```
--> [
        {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},
        {"jsonrpc": "2.0", "method": "notify_hello", "params": [7]},
        {"jsonrpc": "2.0", "method": "subtract", "params": [42,23], "id": "2"},
        {"foo": "boo"},
        {"jsonrpc": "2.0", "method": "foo.get", "params": {"name": "myself"}, "id": "5"},
        {"jsonrpc": "2.0", "method": "get_data", "id": "9"} 
    ]
<-- [
        {"jsonrpc": "2.0", "result": 7, "id": "1"},
        {"jsonrpc": "2.0", "result": 19, "id": "2"},
        {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null},
        {"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found"}, "id": "5"},
        {"jsonrpc": "2.0", "result": ["hello", 5], "id": "9"}
    ]
```
and
```
--> [1]
<-- [
  {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null}
]
```

both cases are handled the way spec defines it.